### PR TITLE
fix and enhancement: Add i3bar option to not unhide on urgency (#4196, #5100)

### LINKED ISSUE #4196: i3bar (with `mode hide`) is raised when moving mouse between monitors and should not be
## I'm submitting a…

<pre>
[x] Bug
[ ] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior

When i3bar is configured with `bar { mode hide }` it should only be visible when pressing the `Mod` key; but when there is more than one monitor it is also raised when moving the mouse between monitors (with `focus_follows_mouse yes`), or when clicking on a window on a different monitor (with `focus_follows_mouse no`).

This requires pressing the `Mod` key to hide it.

This is very disruptive to mental flow because it is typically covering the bottom line of a terminal or the bottom status line of other apps (such as vim), and as many times as it happens I still have a moment of confusion when I start typing a command into vim and don't see it at the bottom of the screen (because it is covered by i3bar).

## Expected Behavior

i3bar should not be raised when moving the mouse between monitors, or clicking on a window to focus it.

When in `mode hide`, i3bar should only be visible when pressing the `Mod` key.

## Reproduction Instructions

Relevant section of i3 config:

    bar {
        status_command i3status
        mode hide
        hidden_state hide
        modifier $mod
        tray_output primary
        font pango: Hack Regular 10
    }

## Environment

Output of `i3 --moreversion 2>&-`:
<pre>
i3 version: 
Binary i3 version:  4.18.1 (2020-04-22) © 2009 Michael Stapelberg and contributors
Running i3 version: 4.18.1 (2020-04-22) (pid 3843)abort…)
Loaded i3 config: /home/user/.config/i3/config (Last modified: Tue 15 Sep 2020 07:57:34 PM +08, 365 seconds ago)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3
</pre>

<details><summary>Config file</summary><pre>

    # This file has been auto-generated by i3-config-wizard(1).
    # It will not be overwritten, so edit it as you like.
    #
    # Should you change your keyboard layout some time, delete
    # this file and re-run i3-config-wizard(1).
    #
    
    # i3 config file (v4)
    #
    # Please see https://i3wm.org/docs/userguide.html for a complete reference!
    
    set $mod Mod4
    
    # swap windows and alt keys
    #exec --no-startup-id setxkbmap -option altwin:swap_alt_win
    exec --no-startup-id setxkbmap -option altwin:swap_lalt_lwin
    
    # Font for window titles. Will also be used by the bar unless a different font
    # is used in the bar {} block below.
    #font pango:monospace 8
    
    # This font is widely installed, provides lots of unicode glyphs, right-to-left
    # text rendering and scalability on retina/hidpi displays (thanks to pango).
    #font pango:DejaVu Sans Mono 8
    font pango:Hack Regular 8
    
    # The combination of xss-lock, nm-applet and pactl is a popular choice, so
    # they are included here as an example. Modify as you see fit.
    
    # xss-lock grabs a logind suspend inhibit lock and will use i3lock to lock the
    # screen before suspend. Use loginctl lock-session to lock your screen.
    exec --no-startup-id xss-lock --transfer-sleep-lock -- i3lock --nofork --color #000000
    
    # NetworkManager is the most popular way to manage wireless networks on Linux,
    # and nm-applet is a desktop environment-independent system tray GUI for it.
    exec --no-startup-id nm-applet
    
    # Use pactl to adjust volume in PulseAudio.
    set $refresh_i3status killall -SIGUSR1 i3status
    bindsym XF86AudioRaiseVolume exec --no-startup-id pactl set-sink-volume @DEFAULT_SINK@ +10% && $refresh_i3status
    bindsym XF86AudioLowerVolume exec --no-startup-id pactl set-sink-volume @DEFAULT_SINK@ -10% && $refresh_i3status
    bindsym XF86AudioMute exec --no-startup-id pactl set-sink-mute @DEFAULT_SINK@ toggle && $refresh_i3status
    bindsym XF86AudioMicMute exec --no-startup-id pactl set-source-mute @DEFAULT_SOURCE@ toggle && $refresh_i3status
    
    # mocp controlls (for music)
    bindsym XF86AudioPrev exec --no-startup-id mocp --previous
    bindsym XF86AudioNext exec --no-startup-id mocp --next
    bindsym XF86AudioPlay exec --no-startup-id mocp --toggle-pause
    bindsym XF86AudioStop exec --no-startup-id mocp --stop
    
    # Use Mouse+$mod to drag floating windows to their wanted position
    floating_modifier $mod
    
    # start a terminal
    #bindsym $mod+Return exec i3-sensible-terminal
    bindsym $mod+Shift+Return exec xfce4-terminal --initial-title Host --dynamic-title-mode after
    #bindsym $mod+Return exec xfce4-terminal
    bindsym $mod+Return exec xfce4-terminal --command ~/bin/devmachine --initial-title Workshop
    
    
    # kill focused window
    bindsym $mod+Shift+colon kill
    
    # start dmenu (a program launcher)
    bindsym $mod+e exec i3-dmenu-desktop
    # There also is the (new) i3-dmenu-desktop which only displays applications
    # shipping a .desktop file. It is a wrapper around dmenu, so you need that
    # installed.
    # bindsym $mod+d exec --no-startup-id i3-dmenu-desktop
    
    # change focus
    bindsym $mod+h focus left
    bindsym $mod+t focus down
    bindsym $mod+n focus up
    bindsym $mod+s focus right
    
    # alternatively, you can use the cursor keys:
    bindsym $mod+Left focus left
    bindsym $mod+Down focus down
    bindsym $mod+Up focus up
    bindsym $mod+Right focus right
    
    # move focused window
    bindsym $mod+Shift+H move left
    bindsym $mod+Shift+T move down
    bindsym $mod+Shift+N move up
    bindsym $mod+Shift+S move right
    
    # alternatively, you can use the cursor keys:
    bindsym $mod+Shift+Left move left
    bindsym $mod+Shift+Down move down
    bindsym $mod+Shift+Up move up
    bindsym $mod+Shift+Right move right
    
    # split in horizontal orientation
    bindsym $mod+d split h
    
    # split in vertical orientation
    bindsym $mod+k split v
    
    # enter fullscreen mode for the focused container
    bindsym $mod+u fullscreen toggle
    
    # change container layout (stacked, tabbed, toggle split)
    bindsym $mod+o layout stacking
    bindsym $mod+comma layout tabbed
    bindsym $mod+period layout toggle split
    
    # toggle tiling / floating
    bindsym $mod+Shift+space floating toggle
    
    # change focus between tiling / floating windows
    bindsym $mod+space focus mode_toggle
    
    # focus the parent container
    bindsym $mod+a focus parent
    
    # focus the child container
    #bindsym $mod+d focus child
    
    # Define names for default workspaces for which we configure key bindings later on.
    # We use variables to avoid repeating the names in multiple places.
    set $ws1 "1"
    set $ws2 "2"
    set $ws3 "3"
    set $ws4 "4"
    set $ws5 "5"
    set $ws6 "6"
    set $ws7 "7"
    set $ws8 "8"
    set $ws9 "9"
    set $ws10 "10"
    
    # switch to workspace
    bindsym $mod+ampersand workspace number $ws1
    bindsym $mod+bracketleft workspace number $ws2
    bindsym $mod+braceleft workspace number $ws3
    bindsym $mod+braceright workspace number $ws4
    bindsym $mod+parenleft workspace number $ws5
    bindsym $mod+equal workspace number $ws6
    bindsym $mod+asterisk workspace number $ws7
    bindsym $mod+parenright workspace number $ws8
    bindsym $mod+plus workspace number $ws9
    bindsym $mod+bracketright workspace number $ws10
    
    # move focused container to workspace
    bindsym $mod+Shift+percent move container to workspace number $ws1; workspace $ws1
    bindsym $mod+Shift+7 move container to workspace number $ws2; workspace $ws2
    bindsym $mod+Shift+5 move container to workspace number $ws3; workspace $ws3
    bindsym $mod+Shift+3 move container to workspace number $ws4; workspace $ws4
    bindsym $mod+Shift+1 move container to workspace number $ws5; workspace $ws5
    bindsym $mod+Shift+9 move container to workspace number $ws6; workspace $ws6
    bindsym $mod+Shift+0 move container to workspace number $ws7; workspace $ws7
    bindsym $mod+Shift+2 move container to workspace number $ws8; workspace $ws8
    bindsym $mod+Shift+4 move container to workspace number $ws9; workspace $ws9
    bindsym $mod+Shift+6 move container to workspace number $ws10; workspace $w10
    
    # reload the configuration file
    bindsym $mod+Shift+J reload
    # restart i3 inplace (preserves your layout/session, can be used to upgrade i3)
    bindsym $mod+Shift+P restart
    # exit i3 (logs you out of your X session)
    bindsym $mod+Shift+greater exec "i3-nagbar -t warning -m 'You pressed the exit shortcut. Do you really want to exit i3? This will end your X session.' -B 'Yes, exit i3' 'i3-msg exit'"
    
    # resize window (you can also use the mouse for that)
    mode "resize" {
            # These bindings trigger as soon as you enter the resize mode
    
            # Pressing left will shrink the window’s width.
            # Pressing right will grow the window’s width.
            # Pressing up will shrink the window’s height.
            # Pressing down will grow the window’s height.
            bindsym h resize shrink width 10 px or 10 ppt
            bindsym t resize grow height 10 px or 10 ppt
            bindsym n resize shrink height 10 px or 10 ppt
            bindsym s resize grow width 10 px or 10 ppt
    
            # same bindings, but for the arrow keys
            bindsym Left resize shrink width 10 px or 10 ppt
            bindsym Down resize grow height 10 px or 10 ppt
            bindsym Up resize shrink height 10 px or 10 ppt
            bindsym Right resize grow width 10 px or 10 ppt
    
            # back to normal: Enter or Escape or $mod+r
            bindsym Return mode "default"
            bindsym Escape mode "default"
            bindsym $mod+p mode "default"
    }
    
    bindsym $mod+p mode "resize"
    
    # Start i3bar to display a workspace bar (plus the system information i3status
    # finds out, if available)
    bar {
            status_command i3status
            output primary
            tray_output primary
            font pango: Hack Regular 10
    }
    bar {
            status_command i3status
            mode hide
            hidden_state hide
            modifier $mod
            tray_output primary
            font pango: Hack Regular 10
    }
    
    # custom mods below
    bindsym $mod+b exec firefox
    bindsym $mod+l exec firefox --private-window
    exec --no-startup-id xrandr --output eDP --scale 0.5x0.5
    
    bindsym $mod+w exec xfce4-terminal --title 'notes' --command 'notes'
    for_window [ title="notes" ] floating enable
    bindsym $mod+v exec xfce4-terminal --title "task"
    for_window [ title="task" ] floating enable
    
    bindsym Shift+$mod+grave move workspace to output right
    bindsym Shift+$mod+8 move workspace to output left
    
    exec --no-startup-id xrdb -merge ~/.Xresources &
    
    # not using urxvt anymore since it doesn't support glyphs well (ie. powerline fonts)
    # exec --no-startup-id urxvtd -q -f -o
    
    set $teams [class="Microsoft Teams" instance="microsoft teams"]
    assign $teams 1
    no_focus $teams
    for_window [class="Microsoft Teams" title="Microsoft Teams Notification" instance="microsoft teams"] floating enable
    
    set $signal [class="Signal" instance="signal" title="Signal" window_role="browser-window"]
    assign $signal 1
    no_focus $signal
    
    set $thunderbird [instance="Mail"]
    assign $thunderbird 1
    no_focus $thunderbird
    
    set $volumepanel [class="Pavucontrol" instance="pavucontrol" title="Volume Control"]
    assign $volumepanel 1
    
    for_window [class="Thunar" instance="Thunar"] floating enable
    
    
    set $thunar [class="Thunar" instance="thunar"]
    for_window $thunar floating enable
    
    set $appfinder [class="Xfce4-appfinder" instance="xfce4-appfinder" title="Application Finder"]
    for_window $appfinder floating enable
    
    workspace_auto_back_and_forth no
    force_display_urgency_hint 500 ms
    
    # read 1 character and mark the current window with this character
    bindsym $mod+m exec i3-input -F 'mark %s' -l 1 -P 'Mark: '
    
    # read 1 character and go to the window with the character
    bindsym $mod+g exec i3-input -F '[con_mark="%s"] focus' -l 1 -P 'Goto: '
    
    # Make the currently focused window a scratchpad
    bindsym $mod+Shift+c move scratchpad
    # Show the first scratchpad window
    bindsym $mod+c scratchpad show
    # Show the sup-mail scratchpad window, if any.
    #bindsym mod4+s [title="^Sup ::"] scratchpad show
    
    exec --no-startup-id xsetroot -solid "#000000"
    
    bindsym $mod+Shift+l exec --no-startup-id i3lock -c "#000000"

</pre>

</details>

<pre>
- Linux Distribution & Version: Fedora 32
- Are you using a compositor (e.g., xcompmgr or compton): No
</pre>

### LINKED ISSUE #5100: Add an option to not unhide i3bar when a window/workspace becomes urgent
## I'm submitting a…

<pre>
[ ] Bug
[x] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior

When in mode hide i3bars become unhidden when a window/workspace becomes urgent.
Hiding them (ie using modifier key) only works temporarily as they unhide themselves again when moving outputs if a window/workspace is still urgent.

## Expected Behavior

The current behaviour is sort-of expected because nothing in the doco promises to not re-unhide i3bars after the user has hidden them, but it's not intuitive, and it's distracting and annoying.

## Environment

Output of `i3 --moreversion 2>&-`:
<pre>
i3 version: 4.20.1 (2021-11-03)
</pre>

This is possibly a duplicate, or related to, issue #4196.

### PR DESCRIPTION
**Pull request by gavine99**
Closes #5100
Possibly closes #4196
Provides an i3bar option to stop unhiding i3bar when a window/workspace becomes urgent.
Includes documentation updates.

### CODE REVIEW

File: docs/ipc [hunk 1]
```diff
@@ -657,6 +657,8 @@ workspace_buttons (boolean)::
 	Display workspace buttons or not? Defaults to true.
 binding_mode_indicator (boolean)::
 	Display the mode indicator or not? Defaults to true.
+dont_unhide_for_urgent (boolean)::
```
Comment (orestisfl) [inline docs/ipc]: A double negative is not a great configuration name.

IMO `show_on_urgent` reads better

File: i3bar/src/xcb.c [hunk 2]
```diff
@@ -2038,7 +2038,9 @@ void draw_bars(bool unhide) {
                     fg_color = colors.urgent_ws_fg;
                     bg_color = colors.urgent_ws_bg;
                     border_color = colors.urgent_ws_border;
-                    unhide = true;
+
+                    if (!config.dont_unhide_for_urgent)
+                        unhide = true;
```
Comment (orestisfl) [inline i3bar/src/xcb.c]:
```suggestion
                    if (!config.dont_unhide_for_urgent) {
                        unhide = true;
                     }
```

**Review by orestisfl — CHANGES_REQUESTED** [review]
Fine with this change overall

File: docs/ipc (hunk 1 above)
Comment (gavine99) [reply docs/ipc]: Agreed. I've pushed a new commit with the user-facing config option named as you suggested. docs updated too.

File: i3bar/src/xcb.c [hunk 3]
```diff
@@ -2038,7 +2038,10 @@ void draw_bars(bool unhide) {
                     fg_color = colors.urgent_ws_fg;
                     bg_color = colors.urgent_ws_bg;
                     border_color = colors.urgent_ws_border;
-                    unhide = true;
+
+                    if (!config.dont_unhide_for_urgent) {
```
Comment (orestisfl) [inline i3bar/src/xcb.c]: I think this would also read better if it was named `show_on_urgent`

File: include/configuration.h [hunk 4]
```diff
@@ -349,6 +349,11 @@ struct Barconfig {
      * but we invert the bool for the same reason as hide_workspace_buttons.*/
     bool hide_binding_mode_indicator;
 
+    /** Don't unhide bar for urgent window/workspaces. Configuration option is
+     * 'dont_unhide_for_urgent yes' but we invert the bool to get the correct 
+     * default when initializing with zero. */
```
Comment (orestisfl) [inline include/configuration.h]: This needs updating as well

**Review by orestisfl — CHANGES_REQUESTED** [review]

**Comment by orestisfl** [comment]
Also, I don't see how this can fix #4196

File: i3bar/src/xcb.c (hunk 3 above)
Comment (gavine99) [reply i3bar/src/xcb.c]: ok. i changed the name throughout and defaulted it's value to true to maintain backwards compatibility

File: include/configuration.h (hunk 4 above)
Comment (gavine99) [reply include/configuration.h]: done

**Comment by gavine99** [comment]
> Also, I don't see how this can fix #4196

I had the same problem as that issue - bar re-showing when focus changed across monitor - and this PR stops it occurring. I believe that issue is actually caused because those people had an urgent window.  Good to get feedback from those users when they get version with this PR I guess.

**Pull request closed without merging**